Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm

```diff
@@ -1552,6 +1554,73 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
     }, webPageIDInMainFrameProcess());
 }
 
+void WebPageProxy::getWebArchiveData(CompletionHandler<void(API::Data*)>&& completionHandler)
+{
+    RefPtr mainFrame = m_mainFrame;
+    if (!mainFrame)
+        return completionHandler(nullptr);
+
+    class WebArchvieCallbackAggregator final : public ThreadSafeRefCounted<WebArchvieCallbackAggregator, WTF::DestructionThread::MainRunLoop> {
+    public:
+        using Callback = CompletionHandler<void(RefPtr<LegacyWebArchive>&&)>;
+        static Ref<WebArchvieCallbackAggregator> create(WebCore::FrameIdentifier rootFrameIdentifier, Callback&& callback)
+        {
+            return adoptRef(*new WebArchvieCallbackAggregator(rootFrameIdentifier, WTFMove(callback)));
+        }
+
+        RefPtr<WebCore::LegacyWebArchive> completeFrameArchive(FrameIdentifier identifier)
+        {
+            RefPtr archive = m_frameArchives.take(identifier);
+            if (archive) {
+                for (auto subframeIdentifier : archive->subframeIdentifiers()) {
+                    if (auto subframeArchive = completeFrameArchive(subframeIdentifier))
+                        archive->appendSubframeArchive(subframeArchive.releaseNonNull());
+                }
+            }
+            return archive;
+        }
+        ~WebArchvieCallbackAggregator()
+        {
+            if (m_callback)
+                m_callback(completeFrameArchive(m_rootFrameIdentifier));
+        }
+
+        void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>&& frameArchives)
+        {
+            for (auto frameIdentifier : copyToVector(frameArchives.keys())) {
+                if (RefPtr frameArchive = frameArchives.take(frameIdentifier))
```

Comment:
It seems rather unnecessary to use take here? Are you trying to avoid a ref-churn?

Reply:
Yes; will change to just iterate directly on map